### DISCUSSION TITLE
Wheel Settings mLongitudinalFriction

### DISCUSSION THREAD

**Discussion by Ono-Sendai** [jrouwe/JoltPhysics#1321] [General]
Hi,
It seems that the wheel friction behaviour has changed in a recent Jolt version (am updating), so am looking at mLongitudinalFriction again.  it's confusing to me though - the commented behaviour is

"///< On the Y-axis: friction in the forward direction of the tire. Friction is normally between 0 (no friction) and 1 (full friction) although friction can be a little bit higher than 1 because of the profile of a tire. On the X-axis: the slip ratio (fraction) defined as (omega_wheel * r_wheel - v_longitudinal) / |v_longitudinal|. You can see slip ratio as the amount the wheel is spinning relative to the floor: 0 means the wheel has full traction and is rolling perfectly in sync with the ground, 1 is for example when the wheel is locked and sliding over the ground."

But shouldn't friction be highest when there is no slipping? e.g. when the slip ratio is zero?
As it is the default curve goes through the origin, e.g. friction is zero when slip ratio is zero:

```
mLongitudinalFriction.Reserve(3);
mLongitudinalFriction.AddPoint(0.0f, 0.0f);
mLongitudinalFriction.AddPoint(0.06f, 1.2f);
mLongitudinalFriction.AddPoint(0.2f, 1.0f);
```
Or am I misunderstanding something?

**Comment by Ono-Sendai** [comment]
Tested with a flat curve for mLongitudinalFriction and it seems to work better, no wheel skidding when accelerating.  So maybe this is a straight-up bug?

**Comment by jrouwe** [comment]
There was a bug in that system before, see:

https://github.com/jrouwe/JoltPhysics/blob/21ba054b4cc657083018193695fecb360139b460/Docs/APIChanges.md?plain=1#L36

There's a workaround to get the old behavior back:

https://github.com/jrouwe/JoltPhysics/blob/a456b244aa2ad2ce0a8124d27823377ed0b1c4b4/Samples/Tests/Vehicle/VehicleConstraintTest.cpp#L156-L164).%20(a456b244aa2ad2ce0a8124d27823377ed0b1c4b4

I agree that the wheels are skidding much more than they used to do, but that's explainable as the friction was 'num solver iterations' times higher than what was configured. I also thought the new behavior felt buggy, so I configured a vehicle in PhysX and compared it with Jolt and the results looked very similar. The setup was to have a vehicle in 1st gear and to floor the gas pedal for 5 seconds:

![image](https://github.com/user-attachments/assets/7982ab91-bad7-464f-8d16-db798587ea78)

![image](https://github.com/user-attachments/assets/dfb792cd-721b-4e15-bd32-45d0f6e470dd)

![image](https://github.com/user-attachments/assets/aff1bd90-f5b2-46d0-9e3f-168fbe52450d)

I don't remember the meaning of all these stats anymore (forgot to write that down), but I think Ffront/Fback is the force of the wheels on the ground, RPM is engine speed, Front/Back are probably wheel rotation speeds, Vel is vehicle velocity, Slip is the slip ratio of the driven wheels.

**Comment by jrouwe** [comment]
And to answer your question:

> But shouldn't friction be highest when there is no slipping? e.g. when the slip ratio is zero?

When slip ratio is zero, the wheel is rotating to match the ground speed perfectly. Friction can only exist if they move at different speeds (otherwise the rubber of the tire does not experience any longitudinal forces from the road), so the curve should be zero for a slip ratio of zero.

**Reply by Ono-Sendai** [reply]
Ah ok I was thinking it was the friction coefficient, instead of the friction force.

**Reply by jrouwe** [reply]
It is a friction coefficient, but the friction coefficient is dependent on the relative motion between the rubber and the road (i.e. non-linear friction).

**Reply by Ono-Sendai** [reply]
What about static friction, you can have friction between objects that aren't moving relative to each other.

**Reply by jrouwe** [reply]
In the tire model that I'm using, there is no friction force if there is no relative movement between the wheel and the ground (`WheelSettingsWV::mLongitudinalFriction.GetValue(0)` == 0), but you can set any curve you want and make that point bigger than zero. I think that will work.

**Reply by Ono-Sendai** [reply]
Let's say you stop your car on a slope.  What is keeping the car from sliding down the slope then in Jolt?

**Reply by jrouwe** [reply]
The solver is iterative so you'd get:
* Car wheel is not rotating, no difference in velocity between floor and wheel, therefore no friction
* Velocity is updated, car about to slide down, now there is a velocity difference between wheel and floor, now there is a max friction impulse it can apply
* Velocity is updated again, difference between wheel velocity and floor changes again, again a new max friction impulse is calculated
* Repeat until max number of iterations
* Update the position of the car
In practice this leads to an equilibrium that leaves the car in place.

**Reply by Ono-Sendai** [reply]
Ok, interesting, thank you.  I see it is a bit more complicated than the mathematically ideal scenario.

**Comment by Ono-Sendai** [comment]
What's the recommended way to modify the friction of a vehicle's tires?
Modify the mLongitudinalFriction curve? Or something else?

**Reply by jrouwe** [reply]
You can:

* Modify `mLongitudinalFriction` and `mLateralFriction` (can do that on the fly)
* Do your own friction calculations with `SetCombineFriction`
* Do your own impulse calculations with `SetTireMaxImpulseCallback`
